File: LS8/cpu.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cpu.h"

// push a value to the CPU stack
void cpu_push(struct cpu *cpu, unsigned char val)
{
    cpu->reg[SP]--;

    cpu->ram[cpu->reg[SP]] = val;
}

// pop a value from the CPU stack
unsigned char cpu_pop(struct cpu *cpu)
{
    unsigned char val = cpu->ram[cpu->reg[SP]];

    cpu->reg[SP]++;

    return val;
}

// helper functions
unsigned char cpu_ram_read(struct cpu *cpu, unsigned char MAR)
{
    // MAR: Memory Address Register, holds the memory address we're reading
    return cpu->ram[MAR];
}

void cpu_ram_write(struct cpu *cpu, unsigned char MAR, unsigned char MDR)
{
    // MDR: Memory Data Register, holds the value to write or the value just read
    cpu->ram[MAR] = MDR;
}

// load the binary bytes from an .ls8 source file into a RAM array
void cpu_load(char *filename, struct cpu *cpu)
{
    // You can use the fopen( ) function to create a new file or to open an existing 
    // file. This call will initialize an object of the type FILE, which contains all 
    // the information necessary to control the stream.
    FILE *fp;
    // allocate 1024 bytes for a string variable named "line"
    char line[1024];
    // check out where ADDR_PROGRAM_Entry comes from
    // answer for ^,  it's in cpu.h, defined as 0xF4
    int address = ADDR_PROGRAM_ENTRY;

    // open the source file
    if ((fp = fopen(filename, "r")) == NULL)
    {
        // error handling
        fprintf(stderr, "Cannot open file %s\n", filename);
        exit(2);
    }

    // read all the lines and store them in RAM
    // fgets = reads a line from the specified stream and stores it 
    // into the string pointed to by str
    while (fgets(line, sizeof(line), fp) != NULL)
    {
        // convert string to a number
        char *endchar;
        unsigned char byte = strtol(line, &endchar, 2);

        // ignore lines from which no numbers were read
        if (endchar == line)
        {
            continue;
        }

        // store it in RAM
        cpu->ram[address++] = byte;
    }
}

void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB)
{
    // variable declarations
    unsigned char *reg = cpu->reg;
    unsigned char valA = reg[regA];
    unsigned char valB = reg[regB];

    switch (op)
    {
        case ALU_MUL:
            valA *= valB;
            break;

        case ALU_ADD:
            valA += valB;
            break;

        case ALU_CMP:
            // compares values
            if (valA == valB) 
            {
                cpu->fl = cpu->fl | FL_EQ; // could be written |=
            }
            else
            {
                cpu->fl = cpu->fl & (~FL_EQ);  // could be written &=
            }
            break;
    }
}

void cpu_run(struct cpu *cpu)
{
    // we create a 'running' variable, and set it to 1, for TRUE
    // it is TRUE until we get an HLT instruction
    int running = 1;

    while (running)
    {
        // 1. Get the value of the current instruction (in address PC).
        // It needs to read the memory address that's stored in register PC, 
        // and store that result in IR, the Instruction Register.
        unsigned char IR = cpu_ram_read(cpu, cpu->pc);
        // read the bytes at PC+1 and PC+2 from RAM into variables operandA 
        // and operandB in case the instruction needs them.
        unsigned char operandA = cpu_ram_read(cpu, cpu->pc+1);
        unsigned char operandB = cpu_ram_read(cpu, cpu->pc+2);

        // True if this instruction might set the counter;
        int instruct_set_pc = (IR >> 4) & 1;


        // 2. switch() over it to decide on a course of action.
        // 3. Do whatever the instruction should do according to the spec.
        switch(IR) {
            case LDI:
                cpu->reg[operandA] = operandB;
                break;

            case PRN:
                printf("%d\n", cpu->reg[operandA]);
                break;

            case MUL:
                alu(cpu, ALU_MUL, operandA, operandB);
                break;

            case ADD:
                alu(cpu, ALU_ADD, operandA, operandB);
                break;

            case PUSH:
                cpu->reg[SP]--;
                cpu_ram_write(cpu, cpu->reg[SP], cpu->reg[operandA]);
                break;
            
            case POP:
                cpu->reg[operandA] = cpu_ram_read(cpu, cpu->reg[SP]);
                cpu->reg[SP]++;
                break;

            case CALL:
                cpu_push(cpu, cpu->pc + 2);
                cpu->pc = cpu->reg[operandA];
                break;

            case RET:
                cpu->pc = cpu_pop(cpu);
                break;

            case JMP:
                cpu->pc = cpu->reg[operandA];
                break;

            case CMP:
                alu(cpu, ALU_CMP, operandA, operandB);
                break;

            case JEQ:
                if (cpu->fl & FL_EQ)
                {
                    cpu->pc = cpu->reg[operandA];
                }
                else
                {
                    instruct_set_pc = 0;
                }
                break;
            
            case JNE:
                if(!(cpu->fl & FL_EQ))
                {
                    cpu->pc = cpu->reg[operandA];
                }
                else
                {
                    instruct_set_pc = 0;   
                }
                break;

            case HLT:
                running = 0;
                break;

            default:
                printf("unkown instruction at %02x: %02x\n", cpu->pc, IR);
                exit(2);
        }
        
        // 4. Move the PC to the next instruction.
        if (!instruct_set_pc)
        {
            // >> splices off 6 characters in IR
            cpu->pc += (IR >> 6) + 1;
        }
    }
}

// Initialize a CPU struct
void cpu_init(struct cpu *cpu)
{
    // init the PC and other special registers
    cpu->pc = 0;
    cpu->fl = 0;

    // zero registers and RAM
    memset(cpu->reg, 0, sizeof cpu->reg);
    memset(cpu->ram, 0, sizeof cpu->ram);

    // init SP
    cpu->reg[SP] = ADDR_EMPTY_STACK;
}

File: ls8/cpu.py
"""
CPU functionality.
"""

import sys

class CPU:
    """
    Main CPU class.
    """
    def __init__(self):
        """
        Construct a new CPU.
        """
        # Opcodes
        self.op = 0
        # RAM memory
        self.ram = [0]*255
        # Registers
        self.reg = [0]*8
        self.reg[7] = 0xF4
        # Program Counter
        self.pc = 0
        # Instruction Register
        self.ir = 0
        # Memory Address Register
        self.mar = 0
        # Memory Data Register
        self.mdr = 0
        # Flag Register
        self.fl = 0
        # Set Running Loop
        self.running = True
        # Stack Pointer
        self.sp = 7

        # in CPU opcodes
        self.LDI = 0b10000010       # LDI regA, index
        self.PRN = 0b01000111       # PRN regA
        self.PRA = 0b01001000       # PRA regA
        self.HLT = 0b00000001       # HLT
        self.ADD = 0b10100000       # ADD regA, regB
        self.MUL = 0b10100010       # MUL regA, regB     
        self.PUSH = 0b01000101      # PUSH regA
        self.POP = 0b01000110       # POP regA
        self.SHL = 0b10101100       # SHL regA, regB
        self.CALL = 0b01010000      # CALL regA
        self.RET = 0b00010001       # RET
     

    def load(self, file_name):
        """
        Load a program into memory.
        """
        address = 0
                
        # with open(file_name[1]) as file:
        with open(file_name) as file:
            
            for line in file.readlines():
                #str = line.strip().partition("#")[0]
                str = line.split("#")[0].strip()
                if len(str) == 0:
                    continue
                
                self.ram_write(address, int(str, 2))
                
                address +=1


    def alu(self, op, oper1, oper2):
        """
        ALU operations.
        """
        if op == "ADD":
            self.reg[oper1] += self.reg[oper2]

        elif op == "MUL":
            self.reg[oper1] *= self.reg[oper2]

        # elif op == "SUB": 
        #     self.reg[oper1] -= self.reg[oper2]

        # elif op == "DIV":
        #     self.reg[oper1] //= self.reg[oper2]

        else:
            raise Exception("Unsupported ALU Operation")

    def trace(self):
        """
        Handy function to print out the CPU state. 
        You might want to call this from run() 
        if you need help debugging.
        """

        print(f"TRACE: %02X | %02X %02X %02X |" % (
            self.pc,
            #self.fl,
            #self.ie,
            self.ram_read(self.pc),
            self.ram_read(self.pc + 1),
            self.ram_read(self.pc + 2)
        ), end='')

        for i in range(8):
            print(" %02X" % self.reg[i], end='')

        print()

    def run(self):
        """
        Run the CPU. ()
        """
        while self.running:
            execute_cmd = self.ram_read(self.pc)
            
            instruction = execute_cmd & 0b00111111  # select opcode and mask
            operand_count = execute_cmd >> 6 
            opcode_size = (operand_count) +1        # shift to right 
            op_position = self.pc
            # operands = (self.ram_read(op_position + i) for i in range(operand_count))
                        
            oper1 = self.ram_read(self.pc+1) #next(operands) 
            oper2 = self.ram_read(self.pc+2) #next(oper1) 

            if execute_cmd == self.LDI: # 0b10000010            
                self.reg[oper1] = oper2

            elif execute_cmd == self.PRN: #0b01000111
                print(self.reg[oper1])

            elif execute_cmd == self.HLT: #0b00000001
                self.running = False

            elif execute_cmd == self.ADD:            
                self.alu("ADD", oper1, oper2)

            elif execute_cmd == self.MUL:
                self.alu("MUL", oper1, oper2)

            elif execute_cmd == self.PUSH:
                # decrement
                self.reg[self.sp] -=1
                # add to stack at memory address assigned by 
                # decremented stack pointer
                self.ram[self.reg[self.sp]] = self.reg[oper1]

            elif execute_cmd == self.POP:
                # copy value at memory address assigned by 
                # stack pointer 
                self.reg[oper1] = self.ram[self.reg[self.sp]]
                # increment
                self.reg[self.sp] +=1

            elif execute_cmd == self.CALL:
                # get the address of the next instruction by adding 2 to 
                # the current instruction
                addr_next_inst = self.pc +2
                # decrement
                self.reg[self.sp] -=1
                # push the address of next instruction onto stack
                # for use in the Return instruction
                self.ram[self.reg[self.sp]] = addr_next_inst

                reg_index = oper1
                addr = self.reg[reg_index]
                self.pc = addr

            elif execute_cmd == self.RET:
                # copy value at memory address assigned by 
                # stack pointer into the pc 
                self.pc = self.ram[self.reg[self.sp]]
                # increment
                self.reg[self.sp] +=1

            else:
                self.trace()
                raise Exception(f'Unrecognized Instruction')

            # increment program counter as determined by opcode size
            # Note: subroutines should not be includes in program counter
            # may need to use a flag and mask to implement
            #if execute_cmd != self.CALL and execute_cmd != self.RET:
            if execute_cmd & 0b00010000 == 0:
                self.pc += opcode_size
                

    def ram_read(self, address):
        """
        Loads registerA with the value at the memory address 
        stored in registerB. This opcode reads from memory.
        """
        return self.ram[address]


    def ram_write(self, address, data):
        """
        Set the value of a register to an integer.
        """
        self.ram[address] = data


    def halt(self):
        self.running = False
        sys.exit(0)
